### PR TITLE
Allows to join multiple lines in visual line mode

### PR DESCRIPTION
**Pull request by shady2k**

### CODE REVIEW

**Comment by WuTheFWasThat** [comment]
do you think the recursive case is important?  I'm worried about having people accidentally trigger very expensive operations, so maybe it should error if things have children.  what do you think?

**Comment by shady2k** [comment]
And what are the other options? Without iterating over child nodes, this operation will look strange.

**Comment by WuTheFWasThat** [comment]
another option is to disallow if anything in the recursive tree is folded.  that way it prevents you from accidentally joining a very large tree (since if it's unfolded it must be small)

**Comment by shady2k** [comment]
Ok, but after joining the folded node will be deleted...

**Comment by WuTheFWasThat** [comment]
ah, no, I meant that no operation will happen in this case, an error message will be displayed to the user

**Comment by shady2k** [comment]
Sounds reasonable, I will change code

**Comment by shady2k** [comment]
@WuTheFWasThat, could you check 3521204683a48257d61358d1b62b3dcbbd118fe4?

File: src/assets/ts/definitions/basics.ts [hunk 1]
```diff
@@ -255,6 +255,38 @@ keyDefinitions.registerAction(new Action(
...
 
+keyDefinitions.registerAction(new Action(
+  'visual-line-join',
+  'Join all selected rows',
+  async function({ session, visual_line }) {
+    if (visual_line == null) {
+      throw new Error('Visual_line mode arguments missing');
+    }
+
+    if (visual_line.num_rows >= 1) {
```
Comment (WuTheFWasThat) [inline src/assets/ts/definitions/basics.ts:266]: should it work even if there's only one row?  then will just join all the children

File: src/assets/ts/session.ts [hunk 2]
```diff
@@ -1335,6 +1335,33 @@ export default class Session extends EventEmitter {
 
     this.emit('scroll', numlines);
   }
+
+  public async getTextRecusive(path: Path) {
+    let result: string[] = [];
+
+    if (await this.document.collapsed(path.row)) {
+      throw new Error('Some blocks is folded!');
```
Comment (WuTheFWasThat) [inline src/assets/ts/session.ts:1343]: nit: "Some blocks are folded!"

File: src/assets/ts/definitions/basics.ts [hunk 3]
```diff
@@ -255,6 +255,38 @@ keyDefinitions.registerAction(new Action(
...
+      try {
+        const resultArr = await Promise.all(
+          visual_line.selected.map(async (path) => {
+            return await session.getTextRecusive(path);
+          })
+        );
+
+        const resultArrClear = resultArr.map(function(x) { return x.replace(/(?:\r)/g, ''); });
```
Comment (WuTheFWasThat) [inline src/assets/ts/definitions/basics.ts:274]: what does this regex do?  could you leave a comment as well?

File: src/assets/ts/session.ts (hunk 2 above)
Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: optional:  I would probably make the error a return value

    let err: string | null = null;

    if (await this.document.collapsed(path.row)) {
      err = 'Some blocks are folded!';
    }

    ...


      for (let [childErr, childResult] of resultChildren) {
        if (childErr !== null) { err = childErr; }
        result.push(childResult);
      }

....
      return err, result.join('\n')

Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: another thing to consider is to make the return value here an array, and do the `result.join('\n')` in `visual-line-join` instead

File: src/assets/ts/definitions/basics.ts (hunk 1 above)
Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:266]: Yes, that was the goal - to join child nodes.

File: src/assets/ts/session.ts (hunk 2 above)
Comment (shady2k) [reply src/assets/ts/session.ts:1343]: > nit: "Some blocks are folded!"
You are right. Sorry, English is not my native language.

Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: no problem!

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:274]: This regex removes \r (Carriage Return) from each line.

File: src/assets/ts/definitions/basics.ts (hunk 1 above)
Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:266]: sorry, I misread (thought it was >1).  Does == 0 happen as well?  this looks good

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:274]: cool, could be good to do in visual-line-join instead as well, what do you think?

File: src/assets/ts/session.ts (hunk 2 above)
Comment (shady2k) [reply src/assets/ts/session.ts:1343]: > optional: I would probably make the error a return value
About this: I can't. I tried many options, and failed to return the value with an error.

for (let [childErr, childResult] of resultChildren) {

Type 'string' is not an array type or a string type. Use compiler option '--downlevelIteration' to allow iterating of iterators.ts(2569)

Can you help with it?

Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: in `tsconfig.json`, add    ```"downlevelIteration": true,``` and then rerun `npm start`, and i think it should work

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:274]: > cool, could be good to do in visual-line-join instead as well, what do you think?

I did not understand, could you explain?

Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:274]: nevermind, seems fine, just leave a comment!  when does the carriage return exist in the text?  (I'm used to unix-like OS so don't know much about it)

Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:274]: If I copy text and then paste text from clipboard on Windows machine, in each line there are \r\n. And it looks like an additional char at the end.
image

Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:274]: cool - could be good to fix that in the pasting function too!  i don't mind leaving it here for now also though

Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:274]: I have not reached the paste function yet :)
There is also an unpleasant bug. When the search mode is open, it is impossible to insert text into the search bar; instead, it is inserted at the cursor position on the previous screen.

File: src/assets/ts/session.ts (hunk 2 above)
Comment (shady2k) [reply src/assets/ts/session.ts:1343]: I can't get it work.

 ```
 public async getTextRecusive(path: Path) {
    let result: string[] = [];
    let err: string | null = null;

    if (await this.document.collapsed(path.row)) {
      err = 'Some blocks are folded!';
    }

    const text = await this.document.getText(path.row);
    result.push(text);

    if (await this.document.hasChildren(path.row)) {
      let children = await this.document.getChildren(path);

      const resultChildren = await Promise.all(
        children.map(async (childrenPath) => {
          return await this.getTextRecusive(childrenPath);
        })
      );

      for (let [childErr, childResult] of resultChildren) {
        if (childErr !== null) { err = childErr; }
        result.push(childResult);
      }
    }

    return err, result.join('\n');
  }
}
```

err from child recursive function does not return.

Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: so it compiles, but `err = childErr` doesn't work correctly?

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:274]: ah you mean pasting into the search bar doesn't work?  good catch

File: src/assets/ts/session.ts (hunk 2 above)
Comment (shady2k) [reply src/assets/ts/session.ts:1343]: Yes, it compiles but in that case childErr is first letter of line, childResult second.

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (shady2k) [reply src/assets/ts/definitions/basics.ts:274]: Yes, I tried to paste text into the search bar and nothing worked. After I closed the search mode, all that I pasted turned out to be in the line where the cursor was.

File: src/assets/ts/session.ts (hunk 2 above)
Comment (WuTheFWasThat) [reply src/assets/ts/session.ts:1343]: you need to also change `visual-line-join`, right?

Comment (shady2k) [reply src/assets/ts/session.ts:1343]: I don't understand how it works. Can I somewhere seethe solution to a similar problem? Google doesn't help.

**Comment by shady2k** [comment]
Please check c9032b9f1a52378b7434ddbc73769c17d3622401

File: src/assets/ts/definitions/basics.ts (hunk 3 above)
Comment (WuTheFWasThat) [reply src/assets/ts/definitions/basics.ts:275]: ok opened this issue https://github.com/WuTheFWasThat/vimflowy/issues/338

File: src/assets/ts/session.ts (hunk 2 above)
Comment (WuTheFWasThat) [reply src/assets/ts/session.ts]: thanks, i'll merge this and look into it myself!

Comment (WuTheFWasThat) [reply src/assets/ts/session.ts]: made it work here, if you're curious: https://github.com/WuTheFWasThat/vimflowy/commit/df958af1c0450408caa37a3b2710b46239cd9f68

Comment (shady2k) [reply src/assets/ts/session.ts]: > made it work here, if you're curious: [df958af](https://github.com/WuTheFWasThat/vimflowy/commit/df958af1c0450408caa37a3b2710b46239cd9f68)

Oh, thanks. I would never code like this.

**Pull request merged**